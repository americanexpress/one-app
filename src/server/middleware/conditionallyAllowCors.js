/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import cors from 'cors';

const devOrigin = /localhost:\d{1,5}/;
const corsOptions = {
  origin: [],
};

export const setCorsOrigins = (newCorsOrigins = []) => {
  corsOptions.origin = process.env.NODE_ENV === 'development'
    ? [...newCorsOrigins, devOrigin]
    : newCorsOrigins;
};

setCorsOrigins();

export default function conditionallyAllowCors(req, res, next) {
  const renderPartialOnly = req.store.getState().getIn(['rendering', 'renderPartialOnly']);

  // The HTML partials will have CORS enabled so they can be loaded client-side
  if (renderPartialOnly) {
    return cors(corsOptions)(req, res, next);
  }

  return cors({ origin: false })(req, res, next);
}
