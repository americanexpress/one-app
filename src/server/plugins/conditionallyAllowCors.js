/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import fastifyCors from '@fastify/cors';

const devOrigin = /localhost:\d{1,5}/;
let corsOrigins = [];

export const setCorsOrigins = (newCorsOrigins = []) => {
  corsOrigins = process.env.NODE_ENV === 'development'
    ? [...newCorsOrigins, devOrigin]
    : newCorsOrigins;
};

setCorsOrigins();

const corsOriginsIncludes = (originHeader) => corsOrigins.some((originStringOrRegExp) => {
  if (originHeader === originStringOrRegExp || originStringOrRegExp === '*') return true;
  if (typeof originStringOrRegExp.test === 'function') return originStringOrRegExp.test(originHeader);
  return false;
});

/**
 * Sets configurable cors when 'renderPartialOnly' is enabled
 * @param {import('fastify').FastifyInstance} fastify app instance
 */
const conditionallyAllowCors = async (fastify) => {
  await fastify.register(fastifyCors, {
    hook: 'preHandler',
    delegator: (req, callback) => {
      const { tracer } = req.openTelemetry();
      const span = tracer.startSpan('conditionallyAllowCors');
      const renderPartialOnly = req.store && req.store.getState().getIn(['rendering', 'renderPartialOnly']);
      // The HTML partials will have CORS enabled so they can be loaded client-side
      const origin = renderPartialOnly ? corsOriginsIncludes(req.headers.origin) : false;
      span.end();
      callback(null, { origin });
    },
  });
};

export default conditionallyAllowCors;
