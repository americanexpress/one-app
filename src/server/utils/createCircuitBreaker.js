/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { monitorEventLoopDelay } from 'perf_hooks';
import CircuitBreaker from 'opossum';
import { getModule } from 'holocron';
import { getServerStateConfig } from './stateConfig';
import { registerCircuitBreaker } from '../metrics/circuit-breaker';

const { rootModuleName } = getServerStateConfig();
let eventLoopDelayThreshold = 250;

export const setEventLoopDelayThreshold = (n) => {
  eventLoopDelayThreshold = Number(n) || 250;
};

export const getEventLoopDelayThreshold = () => eventLoopDelayThreshold;

const options = {
  // Do not use a timeout
  timeout: false,
  // When 10% of requests fail, open the circuit
  errorThresholdPercentage: 10,
  // After 10 seconds, try again.
  resetTimeout: 10e3,
};

const eventLoopDelayHistogram = monitorEventLoopDelay();
eventLoopDelayHistogram.enable();

const checkMaxEventLoopDelay = async () => {
  // Return if root module is not loaded, as that is where threshold is configured
  if (!getModule(rootModuleName)) return;
  // Get event loop delay in milliseconds (from nanoseconds)
  const maxEventLoopDelay = eventLoopDelayHistogram.max / 1e6;
  // Open the circuit if event loop delay is greater than threshold
  if (maxEventLoopDelay > eventLoopDelayThreshold) {
    eventLoopDelayHistogram.reset();
    throw new Error(`Opening circuit, event loop delay (${maxEventLoopDelay}ms) is > eventLoopDelayThreshold (${eventLoopDelayThreshold}ms)`);
  }
};

const createCircuitBreaker = (asyncFuntionThatMightFail) => {
  // asyncFuntionThatMightFail should return false to indicate fallback is not needed
  const breaker = new CircuitBreaker(asyncFuntionThatMightFail, options);
  // Fallback returns true to indicate fallback behavior is needed
  breaker.fallback(() => true);
  // Check the max event loop delay every 500ms
  breaker.healthCheck(checkMaxEventLoopDelay, 500);
  // Track circuit breaker metrics
  registerCircuitBreaker(breaker);
  return breaker;
};

export default createCircuitBreaker;
