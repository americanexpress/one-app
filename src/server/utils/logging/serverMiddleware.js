/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import url from 'url';

import { monkeypatches } from '@americanexpress/lumberjack';

import { startTimer, measureTime } from './timing';
import logger from './logger';

const passThrough = ({ log }) => log;
const tenantUtils = {
  configureRequestLog: passThrough,
};

function getLocale(req) {
  if (req.store) {
    const state = req.store.getState();
    return state.getIn(['intl', 'activeLocale']);
  }
  return undefined;
}

function buildMetaData(req, res) {
  const { headers } = req;
  return {
    method: req.method,
    correlationId: headers['correlation-id'],
    // null to explicitly signal no value, undefined if not expected for every request
    host: headers.host || null,
    referrer: headers.referer || headers.referrer || null,
    userAgent: headers['user-agent'] || null,
    location: res.getHeader('location') || undefined,
    forwarded: headers.forwarded || null,
    forwardedFor: headers['x-forwarded-for'] || null,
    locale: getLocale(req),
  };
}

function logClientRequest(req, res) {
  const fullDuration = measureTime(req);
  const responseDuration = measureTime(res);
  const ttfb = fullDuration - responseDuration;
  const log = {
    type: 'request',
    request: {
      direction: 'in',
      protocol: req.protocol,
      address: {
        uri: url.format({
          protocol: req.protocol,
          port: req.port,
          hostname: req.hostname,
          pathname: req.originalUrl,
        }),
      },
      metaData: buildMetaData(req, res),
      timings: { // https://www.w3.org/TR/navigation-timing/
        // navigationStart: 0,
        // redirectStart: 0,
        // redirectEnd: 0,
        // fetchStart: 0,
        // domainLookupStart: 10,
        // domainLookupEnd: 20,
        // connectStart: 30,
        // secureConnectionStart: 40,
        // connectEnd: 50,
        // requestStart: 60,
        // requestEnd: 70, // optional? not part of the W3C spec
        // responseStart: 80,
        // responseEnd: 90,
        // FIXME: mimic the w3 timing names
        duration: Math.round(fullDuration),
        ttfb: ttfb >= 0 ? Math.round(ttfb) : null,
      },
      statusCode: res.statusCode,
      statusText: res.statusMessage || undefined,
    },
  };

  const configuredLog = tenantUtils.configureRequestLog({ req, res, log });
  logger.info(configuredLog);
}

export const setConfigureRequestLog = (newConfigureRequestLog = passThrough) => {
  tenantUtils.configureRequestLog = newConfigureRequestLog;
};

export default function serverMiddleware(req, res, next) {
  startTimer(req);
  monkeypatches.attachSpy(res, 'writeHead', () => startTimer(res));
  res.on('close', () => logClientRequest(req, res));
  setImmediate(next);
}
