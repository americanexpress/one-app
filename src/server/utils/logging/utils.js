/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { STATUS_CODES } from 'http';

import chalk from 'chalk';

const coloredLevels = {
  error: chalk.red('error'),
  warn: chalk.yellow('warn'),
  log: chalk.blue('log'),
  info: chalk.gray('info'),
};

function printStatusCode(obj) {
  const { request: { statusCode } } = obj;

  if (statusCode === null) {
    return chalk.black.bgRed(statusCode);
  }

  if (statusCode < 300) {
    // 100s, 200s
    return chalk.green(statusCode);
  }
  if (statusCode < 400) {
    // 300s
    return chalk.blue(statusCode);
  }
  if (statusCode < 500) {
    // 400s
    return chalk.yellow(statusCode);
  }
  // 500s
  return chalk.red(statusCode);
}

function printStatusMessage(obj) {
  const { request: { statusText, statusCode } } = obj;

  return statusText || STATUS_CODES[statusCode] || chalk.black.bgRed('Timed Out?');
}

function printDurationTime(obj) {
  const { request: { timings: { duration } } } = obj;
  if (duration < 100) {
    return chalk.green(duration);
  }
  if (duration < 1e3) {
    return chalk.yellow(duration);
  }
  if (duration < 6e3) {
    return chalk.red(duration);
  }
  return chalk.black.bgRed(duration);
}

export {
  coloredLevels,
  printStatusCode,
  printStatusMessage,
  printDurationTime,
};
