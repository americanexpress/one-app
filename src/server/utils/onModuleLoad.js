/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import semver from 'semver';
import { Set as ImmutableSet } from 'immutable';
import { META_DATA_KEY } from '@americanexpress/one-app-bundler';
import { setStateConfig, getClientStateConfig, getServerStateConfig } from './stateConfig';
import { setCorsOrigins } from '../middleware/conditionallyAllowCors';
import readJsonFile from './readJsonFile';
import { extendRestrictedAttributesAllowList, validateSafeRequestRestrictedAttributes } from './safeRequest';
import { setConfigureRequestLog } from './logging/serverMiddleware';
import { setCreateSsrFetch } from './createSsrFetch';
import { setEventLoopDelayThreshold } from './createCircuitBreaker';

// Trim build hash
const { buildVersion } = readJsonFile('../../../.build-meta.json');
const appVersion = buildVersion.slice(0, -9);
let modulesUsingExternals = new ImmutableSet();

const registerModuleUsingExternals = (moduleName) => {
  modulesUsingExternals = modulesUsingExternals.add(moduleName);
};

const clearModulesUsingExternals = () => {
  modulesUsingExternals = modulesUsingExternals.clear();
};

export const getModulesUsingExternals = () => modulesUsingExternals.toJS();

export const setModulesUsingExternals = (moduleNames) => {
  modulesUsingExternals = new ImmutableSet(moduleNames);
};

const logModuleLoad = (moduleName, moduleVersion) => {
  console.info(`Loaded module ${moduleName}@${moduleVersion}`);
};

function validateConfig(configValidators, config) {
  Object.entries(configValidators)
    .forEach(([configKey, {
      client: { validate: validateClient = () => 0 } = {},
      server: { validate: validateServer = () => 0 } = {},
    }]) => {
      validateClient(config.client[configKey]);
      validateServer(config.server[configKey]);
    });
}

export const CONFIGURATION_KEY = 'appConfig';

export default function onModuleLoad({
  module,
  moduleName,
}) {
  const {
    [CONFIGURATION_KEY]: {
      // Root Module Specific
      providedExternals,
      provideStateConfig,
      csp,
      corsOrigins,
      configureRequestLog,
      extendSafeRequestRestrictedAttributes = {},
      createSsrFetch,
      eventLoopDelayThreshold,
      // Child Module Specific
      requiredExternals,
      validateStateConfig,
      requiredSafeRequestRestrictedAttributes = {},
      // Any Module
      appCompatibility,
    } = {},
    [META_DATA_KEY]: metaData,
  } = module;

  if (appCompatibility) {
    if (!semver.satisfies(appVersion, appCompatibility, { includePrerelease: true })) {
      throw new Error(`${moduleName}@${metaData.version} is not compatible with this version of one-app (${appVersion}), it requires ${appCompatibility}.`);
    }
  }

  const serverStateConfig = getServerStateConfig();
  const clientStateConfig = getClientStateConfig();

  if (validateStateConfig) {
    validateConfig(validateStateConfig, {
      server: serverStateConfig,
      client: clientStateConfig,
    });
  }

  if (moduleName === serverStateConfig.rootModuleName) {
    if (!csp) {
      throw new Error('Root module must provide a valid content security policy');
    }
    clearModulesUsingExternals();
    if (provideStateConfig) {
      setStateConfig(provideStateConfig);
    }
    setCorsOrigins(corsOrigins);
    extendRestrictedAttributesAllowList(extendSafeRequestRestrictedAttributes);
    setConfigureRequestLog(configureRequestLog);
    setCreateSsrFetch(createSsrFetch);
    setEventLoopDelayThreshold(eventLoopDelayThreshold);
    logModuleLoad(moduleName, metaData.version);
    return;
  }

  if (providedExternals) {
    console.warn(
      `Module ${moduleName} attempted to provide externals. Only the root module can provide externals.`
    );
  }

  if (requiredExternals) {
    const messages = [];
    const RootModule = global.getTenantRootModule();

    Object.entries(requiredExternals).forEach(([externalName, requestedExternalVersion]) => {
      const providedExternal = RootModule[CONFIGURATION_KEY].providedExternals[externalName];

      if (!providedExternal) {
        messages.push(`External '${externalName}' is required by ${moduleName}, but is not provided by the root module`);
      } else if (!semver.satisfies(providedExternal.version, requestedExternalVersion)) {
        messages.push(`${externalName}@${requestedExternalVersion} is required by ${moduleName}, but the root module provides ${providedExternal.version}`);
      }
    });

    if (messages.length !== 0) {
      throw new Error(messages.join('\n'));
    }

    registerModuleUsingExternals(moduleName);
  }

  validateSafeRequestRestrictedAttributes(requiredSafeRequestRestrictedAttributes);

  logModuleLoad(moduleName, metaData.version);
}
