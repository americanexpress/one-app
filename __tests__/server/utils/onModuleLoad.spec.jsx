/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from 'react';
import { preprocessEnvVar } from '@americanexpress/env-config-utils';
import { META_DATA_KEY } from '@americanexpress/one-app-bundler';
import onModuleLoad, {
  setModulesUsingExternals,
  getModulesUsingExternals,
  CONFIGURATION_KEY,
} from '../../../src/server/utils/onModuleLoad';
// This named export exists only on the mock
// eslint-disable-next-line import/named
import { setStateConfig, getClientStateConfig, getServerStateConfig } from '../../../src/server/utils/stateConfig';
import { setCorsOrigins } from '../../../src/server/middleware/conditionallyAllowCors';
import { extendRestrictedAttributesAllowList, validateSafeRequestRestrictedAttributes } from '../../../src/server/utils/safeRequest';
import { setConfigureRequestLog } from '../../../src/server/utils/logging/serverMiddleware';
import { setCreateSsrFetch } from '../../../src/server/utils/createSsrFetch';
import { getEventLoopDelayThreshold } from '../../../src/server/utils/createCircuitBreaker';

jest.mock('../../../src/server/utils/stateConfig', () => ({
  setStateConfig: jest.fn(),
  getClientStateConfig: jest.fn(),
  getServerStateConfig: jest.fn(() => ({ rootModuleName: 'root-module' })),
}));
jest.mock('@americanexpress/env-config-utils');
jest.mock('../../../src/server/utils/readJsonFile', () => () => ({ buildVersion: '4.43.0-0-38f0178d' }));
jest.mock('../../../src/server/middleware/conditionallyAllowCors', () => ({
  setCorsOrigins: jest.fn(),
}));
jest.mock('../../../src/server/utils/logging/serverMiddleware');
jest.mock('../../../src/server/utils/createSsrFetch');

jest.mock('../../../src/server/utils/safeRequest', () => ({
  extendRestrictedAttributesAllowList: jest.fn(),
  validateSafeRequestRestrictedAttributes: jest.fn(),
}));

const RootModule = () => <h1>Hello, world</h1>;
const csp = "default: 'none'";
describe('onModuleLoad', () => {
  const consoleInfoSpy = jest.spyOn(console, 'info').mockImplementation(() => null);
  const consoleWarnSpy = jest.spyOn(console, 'warn').mockImplementation(() => null);

  beforeEach(() => {
    global.getTenantRootModule = () => RootModule;
    jest.resetAllMocks();
    getServerStateConfig.mockImplementation(() => ({
      rootModuleName: 'some-root',
    }));
    getClientStateConfig.mockImplementation(() => ({
      rootModuleName: 'some-root',
    }));
  });

  afterEach(() => {
    setModulesUsingExternals([]);
  });

  it('does not do anything if there is not an environment variable config', () => {
    expect(
      () => onModuleLoad({ module: { [META_DATA_KEY]: { version: '1.0.0' } } })
    ).not.toThrow();
    expect(preprocessEnvVar).not.toHaveBeenCalled();
  });

  it('does not throw when the one app version is compatible', () => {
    expect(
      () => onModuleLoad({
        module: {
          [CONFIGURATION_KEY]: {
            appCompatibility: '^4.41.0',
          },
          [META_DATA_KEY]: { version: '1.0.1' },
        },
        moduleName: 'some-module',
      })
    ).not.toThrow();
  });

  it('throws when the one app version is incompatible', () => {
    expect(
      () => onModuleLoad({
        module: {
          [CONFIGURATION_KEY]: {
            appCompatibility: '~4.41.0',
          },
          [META_DATA_KEY]: { version: '1.0.2' },
        },
        moduleName: 'some-module',
      })
    ).toThrowErrorMatchingSnapshot();
  });

  it('runs validateConfig if validateStateConfig is supplied with custom validate funcs', () => {
    const validateStateConfig = {
      fakeApiUrl: {
        client: {
          validate: jest.fn(),
        },
        server: {
          validate: jest.fn(),
        },
      },
    };
    onModuleLoad({
      module: {
        [CONFIGURATION_KEY]: {
          validateStateConfig,
        },
        [META_DATA_KEY]: { version: '1.0.3' },
      },
      moduleName: 'some-module',
    });
    expect(validateStateConfig.fakeApiUrl.client.validate).toHaveBeenCalled();
    expect(validateStateConfig.fakeApiUrl.server.validate).toHaveBeenCalled();
  });

  it('runs validateConfig if validateStateConfig supplies a key without validate funcs', () => {
    const validateStateConfig = {
      fakeApiUrl: {},
    };
    expect(() => onModuleLoad({
      module: {
        [CONFIGURATION_KEY]: {
          validateStateConfig,
        },
        [META_DATA_KEY]: { version: '1.0.4' },
      },
      moduleName: 'some-module',
    })).not.toThrow();
  });

  it('calls setStateConfig if provideStateConfig is supplied', () => {
    const provideStateConfig = {
      fakeApiUrl: {
        client: {
          qa: 'http://fake-client-url.example.com',
        },
        server: {
          qa: 'http://fake-server-url.example.com',
        },
      },
    };
    onModuleLoad({
      module: {
        [CONFIGURATION_KEY]: {
          provideStateConfig,
          csp,
        },
        [META_DATA_KEY]: { version: '1.0.5' },
      },
      moduleName: 'some-root',
    });
    expect(setStateConfig).toHaveBeenCalledWith(provideStateConfig);
  });

  it('does not throw if the root module provides the expected versions of required externals', () => {
    RootModule[CONFIGURATION_KEY] = {
      providedExternals: {
        'dep-a': { version: '2.1.0', module: () => 0 },
        'dep-b': { version: '5.8.6', module: () => 0 },
        'dep-c': { version: '3.0.10', module: () => 0 },
      },
    };
    const configuration = {
      requiredExternals: {
        'dep-a': '^2.0.0',
        'dep-b': '~5.8.0',
        'dep-c': '>1.0.0',
      },
    };
    expect(() => onModuleLoad({ module: { [CONFIGURATION_KEY]: configuration, [META_DATA_KEY]: { version: '1.0.7' } }, moduleName: 'my-awesome-module' })).not.toThrow();
  });

  it('warns if a module that isn\'t the root module attempts to provide externals', () => {
    const configuration = {
      providedExternals: {
        'dep-b': {
          version: '1.0.0',
          module: () => null,
        },
      },

    };
    onModuleLoad({ module: { [CONFIGURATION_KEY]: configuration, [META_DATA_KEY]: { version: '1.0.8' } }, moduleName: 'my-awesome-module' });
    expect(consoleWarnSpy).toHaveBeenCalledTimes(1);
  });

  it('throws if the root module does not provide the expected external', () => {
    RootModule[CONFIGURATION_KEY] = {
      providedExternals: {
        'dep-a': { version: '2.1.0', module: () => 0 },
      },
    };
    const configuration = {
      requiredExternals: {
        'dep-b': '^2.0.0',
      },
    };
    expect(() => onModuleLoad({ module: { [CONFIGURATION_KEY]: configuration, [META_DATA_KEY]: { version: '1.0.9' } }, moduleName: 'my-awesome-module' })).toThrowErrorMatchingSnapshot();
  });

  it('throws if the root module provides an incompatible version of a required external', () => {
    RootModule[CONFIGURATION_KEY].providedExternals = {
      'dep-a': { version: '2.1.0', module: () => 0 },
    };
    const configuration = {
      requiredExternals: {
        'dep-a': '^2.1.1',
      },
    };
    expect(() => onModuleLoad({ module: { [CONFIGURATION_KEY]: configuration, [META_DATA_KEY]: { version: '1.0.10' } }, moduleName: 'my-awesome-module' })).toThrowErrorMatchingSnapshot();
  });

  it('includes messages about all missing or incompatible externals', () => {
    RootModule[CONFIGURATION_KEY] = {
      providedExternals: {
        'dep-a': { version: '3.2.0', module: () => 0 },
        'dep-b': { version: '5.9.6', module: () => 0 },
        'dep-c': { version: '3.0.10', module: () => 0 },
      },
    };
    const configuration = {
      requiredExternals: {
        'dep-a': '^2.0.0',
        'dep-b': '~5.8.0',
        'dep-c': '>1.0.0',
        'dep-d': '^3.1.2',
      },
    };
    expect(() => onModuleLoad({ module: { [CONFIGURATION_KEY]: configuration, [META_DATA_KEY]: { version: '1.0.11' } }, moduleName: 'my-awesome-module' })).toThrowErrorMatchingSnapshot();
  });

  it('keeps track of the externals a module is using', () => {
    RootModule[CONFIGURATION_KEY] = {
      providedExternals: {
        'dep-a': { version: '2.1.0', module: () => 0 },
        'dep-b': { version: '5.8.6', module: () => 0 },
        'dep-c': { version: '3.0.10', module: () => 0 },
      },
    };
    const configuration = {
      requiredExternals: {
        'dep-a': '^2.0.0',
        'dep-b': '~5.8.0',
      },
    };
    expect(getModulesUsingExternals()).toEqual([]);
    onModuleLoad({ module: { [CONFIGURATION_KEY]: configuration, [META_DATA_KEY]: { version: '1.0.12' } }, moduleName: 'my-awesome-module' });
    expect(getModulesUsingExternals()).toEqual(['my-awesome-module']);
  });

  it('clears the modules using externals when loading the root module', () => {
    const modulesUsingExternals = ['a', 'b', 'c'];
    setModulesUsingExternals(modulesUsingExternals);
    expect(getModulesUsingExternals()).toEqual(modulesUsingExternals);
    onModuleLoad({ module: { [CONFIGURATION_KEY]: { csp }, [META_DATA_KEY]: { version: '1.0.13' } }, moduleName: 'some-root' });
    expect(getModulesUsingExternals()).toEqual([]);
  });

  it('validates csp added to the root module', () => {
    const callOnModuleLoad = () => (onModuleLoad({
      module: {},
      moduleName: 'some-root',
    }));

    expect(callOnModuleLoad).toThrowErrorMatchingSnapshot();
  });

  it('updates createSsrFetch when added on the root module', () => {
    const fakeCreateSsrFetch = jest.fn();
    onModuleLoad({
      module: {
        [CONFIGURATION_KEY]: {
          csp,
          createSsrFetch: fakeCreateSsrFetch,
        },
        [META_DATA_KEY]: { version: '1.0.14' },
      },
      moduleName: 'some-root',
    });

    expect(setCreateSsrFetch).toHaveBeenCalledWith(fakeCreateSsrFetch);
  });

  it('sets CORS origins from the root module', () => {
    const corsOrigins = ['example.com'];
    onModuleLoad({
      module: {
        [CONFIGURATION_KEY]: {
          csp,
          corsOrigins,
        },
        [META_DATA_KEY]: { version: '1.0.14' },
      },
      moduleName: 'some-root',
    });
    expect(setCorsOrigins).toHaveBeenCalledWith(corsOrigins);
  });

  it('sets the event loop lag threshold from the root module', () => {
    const eventLoopDelayThreshold = 50;
    expect(getEventLoopDelayThreshold()).not.toBe(eventLoopDelayThreshold);
    onModuleLoad({
      module: {
        [CONFIGURATION_KEY]: {
          csp,
          eventLoopDelayThreshold,
        },
        [META_DATA_KEY]: { version: '1.0.14' },
      },
      moduleName: 'some-root',
    });
    expect(getEventLoopDelayThreshold()).toBe(eventLoopDelayThreshold);
  });

  it('logs when the root module is loaded', () => {
    onModuleLoad({
      module: { [CONFIGURATION_KEY]: { csp }, [META_DATA_KEY]: { version: '1.0.15' } },
      moduleName: 'some-root',
    });
    expect(consoleInfoSpy).toHaveBeenCalledTimes(1);
    expect(consoleInfoSpy).toHaveBeenCalledWith('Loaded module some-root@1.0.15');
  });

  it('logs when other modules are loaded', () => {
    onModuleLoad({
      module: { [META_DATA_KEY]: { version: '1.0.16' } },
      moduleName: 'not-the-root-module',
    });
    expect(consoleInfoSpy).toHaveBeenCalledTimes(1);
    expect(consoleInfoSpy).toHaveBeenCalledWith('Loaded module not-the-root-module@1.0.16');
  });

  it('updates allowed safeRequest values from the root module', () => {
    onModuleLoad({
      module: {
        [CONFIGURATION_KEY]: {
          extendSafeRequestRestrictedAttributes: { cookies: ['some-cookie'] },
          csp,
        },
        [META_DATA_KEY]: { version: '1.0.5' },
      },
      moduleName: 'some-root',
    });
    expect(extendRestrictedAttributesAllowList).toHaveBeenCalledWith({
      cookies: ['some-cookie'],
    });
  });

  it('validates safeRequest allow list', () => {
    onModuleLoad({
      module: {
        [CONFIGURATION_KEY]: {
          requiredSafeRequestRestrictedAttributes: { cookies: ['some-cookie'] },
        },
        [META_DATA_KEY]: { version: '1.0.5' },
      },
      moduleName: 'random-root',
    });

    expect(validateSafeRequestRestrictedAttributes).toHaveBeenCalledWith({
      cookies: ['some-cookie'],
    });
  });

  it('sets configureRequestLog when given on the root module', () => {
    const configureRequestLog = jest.fn();
    onModuleLoad({
      module: {
        [CONFIGURATION_KEY]: {
          csp,
          configureRequestLog,
        },
        [META_DATA_KEY]: { version: '1.0.14' },
      },
      moduleName: 'some-root',
    });

    expect(setConfigureRequestLog).toHaveBeenCalledWith(configureRequestLog);
  });
});
