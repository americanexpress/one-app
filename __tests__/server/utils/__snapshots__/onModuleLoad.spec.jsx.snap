// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`onModuleLoad includes messages about all missing or incompatible externals 1`] = `
"dep-a@^2.0.0 is required by my-awesome-module, but tenant root module provides 3.2.0
dep-b@~5.8.0 is required by my-awesome-module, but tenant root module provides 5.9.6
External 'dep-d' is required by my-awesome-module, but is not provided by tenant root module"
`;

exports[`onModuleLoad throws if the tenant root module does not provide the expected external 1`] = `"External 'dep-b' is required by my-awesome-module, but is not provided by tenant root module"`;

exports[`onModuleLoad throws if the tenant root module provides an incompatible version of a required external 1`] = `"dep-a@^2.1.1 is required by my-awesome-module, but tenant root module provides 2.1.0"`;

exports[`onModuleLoad throws when the one app version is incompatible 1`] = `"some-module@1.0.2 is not compatible with this version of one-app (4.43.0-0), it requires ~4.41.0."`;

exports[`onModuleLoad validates csp added to tenant module 1`] = `"Root module must provide a valid content security policy"`;
